[BD-14] [SE-2931] User Access Permissions added for Library.

Kelketek opened the pull request:
This PR adds the ability to manage user permissions for librariesv2.

**Discussions**: https://docs.google.com/document/d/1hRrSmWWlGHhi-bh9AIlpSuLtJFnn2GpavNe8OQxthbE/edit

**Dependencies**: None

**Screenshots**: 
![Screen Shot 2020-09-04 at 10 42 20 AM](https://user-images.githubusercontent.com/1099483/92257907-97107080-ee9b-11ea-9ae7-cde291923ff5.png)
![Screen Shot 2020-09-04 at 10 42 41 AM](https://user-images.githubusercontent.com/1099483/92257919-9d9ee800-ee9b-11ea-83aa-9ff95f466164.png)
![Screen Shot 2020-09-04 at 10 42 57 AM](https://user-images.githubusercontent.com/1099483/92257925-a1326f00-ee9b-11ea-820a-566dc8826040.png)

**Merge deadline**: This PR is for code review only and is not to be merged.

**Testing instructions**:

1. Use the `edx-platform` branch `open-craft:fox/se-2931-user-access-adjustments` on the backend.
1. Follow the directions in https://github.com/edx/frontend-app-library-authoring/pull/4 up until item 15.
1. Create a new library of the 'complex' type.
1. In the top menu, there should be a 'User Access' link. Click it to visit the user access page.
1. Create and remove users, elevate their access, and depriv them. Verify the form resets upon add and that the user is added to the list.
1. Put bogus users and data into the email field to verify errors are handled well.
1. Delete users. Note, you should be able to delete yourself as a staff member, and still readd yourself, despite not having labeled privileges.
1. Verify that the interface does not let you remove or depriv yourself if there is no secondary admin.
1. Elevate a user who is not marked is_staff or is_superuser to 'staff' within the interface (that is, library author-- not marking them is_staff in the admin), and have them log in. Verify that the access control buttons are not present.


**Author notes and concerns**:

1. This is the largest amount of React code I've had to make so far, and the first time I've done real full component dev-- previously I modified other components. This is also my first time with Redux. I'd like feedback on how well I'm using those tools.
2. This task initially had an implementation similar to other pages currently in the repository. However, after speaking with @bradenmacdonald , he suggested using [the container pattern](https://www.freecodecamp.org/news/react-superpowers-container-pattern-20d664bdae65/) so that unit tests may be more easily written next sprint. This is my first time using this pattern and so I'd like his review to make sure I'm using/understanding it well.
3. It's been a while since I've done anything a11y related-- we're using paragon, which I believe is able to automate much of it, but I'd like to make sure I understand if I'm assuming its capabilities to be greater than they are in the code I'm writing, since I haven't immediately thought of any markup I'd have to customize here.

**Reviewers**
- [ ] @mavidser 
- [x] @bradenmacdonald 
- [x] edX reviewer[s] TBD (Still need to have an arch level review here, yes?)

bradenmacdonald commented on src/library-header/LibraryHeader.jsx:
```diff
@@ -0,0 +1,108 @@
+                    </a>
+                  </h3>
+                </li>
+                <li className="nav-item nav-account-user">
```
This part (user avatar, logout link, etc.) feels like it should exist in a shared component that someone has made that can be re-used?

bradenmacdonald replied:
And it would be good if there was a way to avoid needing to ship the Studio logo separately in this frontend - does every LMS frontend ship the LMS logo?

kdmccormick replied:
Our MFE header story is in need of improvement.

There exists [a header component](https://github.com/edx/frontend-component-header), which is suitable for many MFEs, but not all of them. That's where the logo comes from for something like [account.edx.org](https://account.edx.org). It's not currently customizable, though, so frontends like Learning that want a more context-specific header end up rolling their own header and shipping their own copy of the logo file.

I have plans to talk to @davidjoy and others to see whether we can beef up the header component so that it can be used in this project. But in the mean time, and especially for the purposes of a prototype, feel free to roll your header component, duplicating what you need to.

arbrandes replied:
What Kevin said. :)

This is the first MFE that needs to reproduce the Studio header, so I'm having to reinvent the wheel, a bit.  At least for now.

bradenmacdonald commented on src/library-header/index.scss:
```diff
@@ -0,0 +1,180 @@
+      @extend %t-action4;
+
+      display: inline-block;
+      vertical-align: middle;
```
Both Bootstrap and edX are moving toward utility-first CSS: https://frontstuff.io/in-defense-of-utility-first-css

Instead of writing specific classes like `library-header`, you put _utilitites_ in the HTML like `<div className="header large-text bottom-border sticky-on-scroll">`, and you can re-use them and recombine them in infinite ways, making CSS much simpler and predictable, while also placing the most useful information directly inline with the HTML.

Could you read the link I posted about that pattern and keep it in mind while you're writing future components?

arbrandes replied:
Agreed, completely.

The CSS you see in this PR is admittedly a bit of a mess, but this is mostly because I opted to use the SASS from Studio as unmodified as possible, as a time-saving measure.  Nevertheless, the work done in creating the subset in `vendor/studio.sass` will make it easier to create decent utilitarian CSS when iterating later.

bradenmacdonald commented on src/library-list/LibraryItem.jsx:
```diff
@@ -0,0 +1,64 @@
+  children: PropTypes.node.isRequired,
+};
+
+const LibraryItem = ({intl, library}) => (
```
Please add a JSDoc comment - it's not immediately obvious from the name what a `LibraryItem` is. I thought it might mean an XBlock in a Library.

arbrandes replied:
+1

bradenmacdonald commented on src/library-common/data/constants.js:
```diff
@@ -0,0 +1,5 @@
+export const LIBRARY_TYPES = {
+  COMPLEX: 'complex',
+  LEGACY: 'legacy',
```
Please add comments defining these - legacy = modulestore/v1, COMPLEX = blockstore/v2

arbrandes replied:
+1

bradenmacdonald commented on src/library-list/data/service.js:
```diff
@@ -0,0 +1,52 @@
+  const base_url = getConfig().CMS_BASE_URL;
+
+  /* Fetch the modulestore libraries, normalizing properties to conform to the v2 API. */
+  const v1Response = await client.get(`${base_url}/library/`);
```
How come here you're using async/await/promises, but in sagas.js there is the unwieldy generators/yield syntax?

arbrandes replied:
I'm not a fan of this inconsistency either, but that's just the way sagas are implemented.  It's the same way in most of the other MFEs I've checked, too.  (Except [frontend-app-learning](https://github.com/edx/frontend-app-learning), but they're testing a different redux flow, there.)

Aside from following the trend in MFEs, I like how sagas reduce a bit of the redux boilerplate.  They also supposedly make it easier to unit test redux code, but I haven't gotten this far, yet.

bradenmacdonald commented on src/library-list/data/service.js:
```diff
@@ -0,0 +1,52 @@
+  }));
+
+  /* Fetch the blockstore libraries. */
+  const v2Response = await client.get(`${base_url}/api/libraries/v2/`);
```
Would be better to fetch this in parallel with the v1 response.

arbrandes replied:
Excellent point.  Done.

mavidser commented on src/library-common/data/utils.js:
```diff
@@ -0,0 +1,39 @@
+
+/* Add a "url" property to the library object, set to the proper library URL
+ * according to its type. */
+export const setLibraryUrl = (library) => {
```
`setLibraryUrl` -> `defineLibraryUrl` or `initLibraryUrl` or something more appropriate for a function with no "value" parameter?

arbrandes replied:
Good point.  +1.

mavidser commented on src/library-common/LoadingPage.jsx:
```diff
@@ -0,0 +1,37 @@
+}
+
+PageLoading.propTypes = {
+  srMessage: PropTypes.string.isRequired,
```
`srMessage` -> `loadingMessage`

arbrandes replied:
+1

mavidser commented on src/library-detail/LibraryDetailPage.jsx:
```diff
@@ -0,0 +1,135 @@
+  }
+
+  render() {
+    const { error, loaded, loading } = this.props;
+
+    let content;
+    if (error) {
+      content = this.renderError();
+    } else if (loading) {
+      content = this.renderLoading();
+    } else if (loaded) {
+      content = this.renderContent();
+    }
```
Use a `loadingState` property to define these mutually exclusive states in a single property instead of separate boolean properties to prevent potential inconsistencies

mavidser replied:
This would also simplify the reducers in the rest of the code

arbrandes replied:
I take your point with respect to conflating `loading` and `loaded`.  However, this will require some refactoring later on anyway: I inherited this pattern from other MFEs, but I think the loading spinner can be better positioned in the skeleton of the page, instead of replacing the whole page.  Let's revisit this when after the MVP is delivered.

mavidser commented on src/library-list/library-form/LibraryForm.jsx:
```diff
@@ -0,0 +1,232 @@
+    };
+  }
+
+  onChangeTitle = (event) => {
+    const data = {...this.state.data, title: event.target.value};
+    this.setState({data});
+  }
+
+  onChangeOrg = (event) => {
+    const data = {...this.state.data, org: event.target.value};
+    this.setState({data});
+  }
+
+  onChangeSlug = (event) => {
+    const data = {...this.state.data, slug: event.target.value};
+    this.setState({data});
+  }
+
+  onChangeType = (event) => {
+    const data = {...this.state.data, type: event.target.value};
+    this.setState({data});
```
can probably use a single `onValueChange` for all events?

arbrandes replied:
Good call.  Done.

mavidser commented:
Haven't gone through complete testing yet, but should we be including a way back to studio in the new MFE? When I click "Libraries" in studio, there seems to be no quick way of going back, with those tabs dissapearing

arbrandes commented:
@kdmccormick, if you can give this (together with the other PRs I pinged you on) a whirl as instructed above while we get the rest of the MVP ready, it would be great.  Thanks!

kdmccormick commented:
@arbrandes On it!

davidjoy commented:
Hi all - 

As mentioned the other day in the #content-libraries slack channel (https://openedx.slack.com/archives/C014K5RST46/p1593027566006300), I'd strongly suggest avoiding using redux-saga unless you have need of its particular features for handling super-complex asynchronous logic.  If we're just using it as a layer to handle side effects in redux actions, redux-thunk is a much, much simpler and more understandable tool.  

In the MFEs that use redux-saga we've found 1) many developers can't figure out what's going on, 2) testing and maintaining them has become a real challenge, and 3) not even browsers do a great job of walking through sagas in the debugger.  Put another way: generator functions are confusing for everybody.

Furthermore, `@reduxjs/toolkit` is an official wrapper for redux.  It's a great way to reduce redux-related boilerplate code, and also comes bundled with `redux-thunk` and `reselect`, a library for creating memoized selectors.  Very helpful for making redux much less verbose.

Happy to talk about any of the above.

arbrandes commented:
ACKing all the feedback so far: thanks a bunch, @bradenmacdonald, @kdmccormick, @mavidser, and @davidjoy.  I'll address everything as best I can over the next week.

@kdmccormick, I'm fine with merging + follow-up PRs.  In practice, though, the first thing is either creating the edX-owned repository, or transfering ownership of this one.  What do you think is best?  In any case, I'm in agreement with your take on what should be tackled afterwards.

@davidjoy, as I mentioned in a comment above, I went with redux-saga for the reduction in boilerplate.  Having worked with [Tornado generators](https://www.tornadoweb.org/en/stable/gen.html) in the past, I didn't find it conceptually difficult.  That said, I'm in no position to defend redux-saga or even JS generator functions, this being the first time I'm using them. :)  So what I'll do is take your suggestion to convert this to `@reduxjs/toolkit`.  (This is not a bad time to time to do it, as I haven't really tackled unit tests, yet.)

kdmccormick commented:
@arbrandes I just got https://github.com/edx/frontend-app-library-authoring created. Feel free to either merge this PR and then pull in the changes over at the edX fork, or open a new PR against the edX fork. Whatever's easier for you.

Commit:
Fix/improve several UI issues.

The alerts needed upgrading to conform to paragon version 10.

Fix uploaded assets not showing

Fix a bug where assets would not show immediately after uploading them.

Also fix the delete button layout; this came about due to the paragon
upgrade to v10.

Fix edit block buttons

Another fix necessitated by the paragon upgrade.

Update header to better conform to Studio design

This fixes the account dropdown in the header so that it matches the
Studio version more closely.

Commit:
Enable filtering on the library list page

This introduces filtering in the library list page.

Commit:
Adds the Edit Library page

Introduces a page for editing library details.  Currently supports
editing the title, description, allow_public_learning, and
allow_public_read fields.

Adds a "Settings" header menu with a "Details" item to reach this page.

Further fixes the header and body layout to better match Studio.

Kelketek committed:
User Access Permissions added for Library.

Kelketek committed:
Address notes on user access.

bradenmacdonald commented on src/library-authoring/library-access/LibraryAccessForm.jsx:
```diff
@@ -0,0 +1,180 @@
+        <Card>
+          <Card.Body>
+            <div className="form-create">
+              <Card.Title>Grant Access to This Library</Card.Title>
```
i18n needed

bradenmacdonald commented on src/library-authoring/library-access/LibraryAccessForm.jsx:
```diff
@@ -0,0 +1,180 @@
+              labels={{
+                disabled: intl.formatMessage(messages['library.access.form.button.submit']),
+                enabled: intl.formatMessage(messages['library.access.form.button.submit']),
+                pending: intl.formatMessage(messages['library.access.form.button.submitting']),
```
Submit/Submit/Submitting seems like it should be a standard i18n message ID used throughout the whole app, or even throughout all Open edX frontends, not something specific to `library.access.form` ?

bradenmacdonald commented on src/library-authoring/library-access/LibraryAccessForm.jsx:
```diff
@@ -0,0 +1,180 @@
+              disabledStates={['disabled', 'pending']}
+              className="action-button"
+            />
+            <Button size="lg" variant="secondary" className="mx-3 action-button" onClick={() => setShowAdd(false)}>Cancel</Button>
```
i18n

bradenmacdonald commented on src/library-authoring/library-access/LibraryAccessForm.jsx:
```diff
@@ -0,0 +1,180 @@
+    props.addUser({ libraryId: props.library.id, data }).then(() => {
+      setData(initialFormState());
+    });
+  };
```
I think these onFoo handler should all be wrapped with [`useCallback`](https://reactjs.org/docs/hooks-reference.html#usecallback); otherwise, the reference value of `onSubmit` changes each time this function is called, resulting in the `onSubmit` prop of `<LibraryAccessForm>` changing each time (to a different but identical function), causing unnecessary renders. Not a big deal in this sort of small UI but it's best to adopt good practices all the time.

bradenmacdonald commented on src/library-authoring/library-access/LibraryAccessForm.jsx:
```diff
@@ -0,0 +1,180 @@
+LibraryAccessForm.propTypes = {
+  intl: intlShape.isRequired,
+  // eslint-disable-next-line react/forbid-prop-types
+  data: PropTypes.object.isRequired,
```
Instead of disabling [this `react/forbid-prop-types` eslint rule](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/forbid-prop-types.md), it would be better to specify `data` using `PropTypes.shape` e.g. 

```
PropTypes.shape({ email: PropTypes.string, }).isRequired,
```

bradenmacdonald commented on src/library-authoring/library-access/LibraryAccessForm.jsx:
```diff
@@ -0,0 +1,180 @@
+  addUser: PropTypes.func.isRequired,
+  setShowAdd: PropTypes.func.isRequired,
+  submissionStatus: PropTypes.oneOf(Object.values(SUBMISSION_STATUS)).isRequired,
+  errorFields: PropTypes.object, // eslint-disable-line react/forbid-prop-types
```
Same here about `PropTypes.object` -> `PropTypes.shape`

bradenmacdonald commented on src/library-authoring/library-access/UserAccessContainer.jsx:
```diff
@@ -0,0 +1,183 @@
+                    </div>
+                    )}
+                  buttons={[
+                    <Button onClick={() => removeAccess()}>Yes.</Button>,
```
i18n

bradenmacdonald commented on src/library-authoring/library-access/data/shapes.js:
```diff
@@ -0,0 +1,10 @@
+
+// eslint-disable-next-line import/prefer-default-export
+export const libraryUserShape = PropTypes.shape({
+  user_id: PropTypes.number.isRequired,
```
Per discussion on the backend PR, we'll need to remove `user_id`

bradenmacdonald commented on src/library-authoring/library-access/index.scss:
```diff
@@ -0,0 +1,34 @@
+  padding: .25rem .75rem .25rem .75rem;
+  color: white;
+  font-size: 70%;
+  text-transform: uppercase;
```
This seems like a good place to refactor this CSS to be more [utility-first](https://tailwindcss.com/docs/utility-first).

In fact, there should already be some utility classes you can use so the HTML would become 

```
<div className={`position-absolute ml-1 px-1 py-3 text-uppercase perm-badge ${user.access_level}`}>
```
and you can then remove most of those from `.perm-badge`, ideally removing that class altogether.

Perhaps the [`badge` components](https://getbootstrap.com/docs/4.0/components/badge/) also have what you need already?

bradenmacdonald commented on src/library-authoring/library-access/index.scss:
```diff
@@ -0,0 +1,34 @@
+
+.btn.action-button {
+  font-weight: bold;
+  text-transform: uppercase;
```
Utility first: just put Bootstrap's standard `className="font-weight-bold text-uppercase"` on the button itself; no need for new CSS classes here.

bradenmacdonald commented on src/library-authoring/library-access/LibraryAccessForm.jsx:
```diff
@@ -0,0 +1,180 @@
+      state = 'disabled';
+    }
+    return state;
+  };
```
This `getSubmitButtonState` does not need to be a function since it doesn't take any arguments. It should just be a string prop, and here you can just calculate its value inline in this `LibraryAccessFormContainer` function and pass the result as a string prop into `LibraryAccessForm`.

If the calculation was in any way expensive, you could wrap it in `useMemo` but it's just a trivial `if` statement so it's fine to run every time this `LibraryAccessFormContainer` function runs. That's better than what you have now, where every time this `LibraryAccessFormContainer` function runs, `getSubmitButtonState` becomes a new reference (to an identical function), causing `LibraryAccessForm` to render again because its `getSubmitButtonState` has technically changed to a new reference.

bradenmacdonald commented on src/library-authoring/library-access/LibraryAccessForm.jsx:
```diff
@@ -0,0 +1,180 @@
+  access_level: LIBRARY_ACCESS.USER,
+});
+
+const LibraryAccessFormContainer = (
```
Nit/optional: I recommend putting a JSDoc-style comment above every React component, with a couple lines explaining what it does and what other components it interacts with.

e.g. 
```
/**
 * LibraryAccessFormContainer:
 * Widget for editing the list of users with access to this content library.
 * This wraps LibraryAccessForm and handles API calls to fetch/write data, as well as displaying 
 * a loading message during initial loading. This separation allows LibraryAccessForm to be tested
 * in unit tests without needing to mock the API.
 */

bradenmacdonald commented on src/library-authoring/library-access/data/api.js:
```diff
@@ -0,0 +1,68 @@
+  return response.data;
+}
+
+export async function setUserAccessLevel({ libraryId, user, level }) {
```
Nit: I don't like the name `user` as a parameter, especially in plain JS with no type information. Is it a user object? an ID? an email? Using `userId` or `userObj` would make it more clear. Though as mentioned in the other PR, we'll likely need to change to just `username`.

Kelketek committed:
Remove references to user ID

Kelketek committed:
Fix some notes.

Kelketek commented:
@bradenmacdonald I've pushed some changes that should address your notes, save for one of the nits.

bradenmacdonald commented:
Great, thanks @Kelketek ! That minor feedback about React was all I had for now. I'll let @mavidser do the main review+test of this once it's unblocked and approve this PR. I assume that Kyle will then do a quick review+merge it for you. If you'd like any further review from me, let me know.

kdmccormick commented on src/library-authoring/studio-header/StudioHeader.jsx:
```diff
@@ -50,6 +50,7 @@ const StudioHeader = ({ intl, library }) => {
                     </Dropdown.Toggle>
                     <Dropdown.Menu className="p-4 mt-1 fade">
                       <Dropdown.Item className="p-0" as={Link} to={ROUTES.Detail.EDIT_SLUG(library.id)}>{intl.formatMessage(messages['library.header.settings.details'])}</Dropdown.Item>
+                      <Dropdown.Item className="p-0" as={Link} to={ROUTES.Detail.ACCESS_SLUG(library.id)}>{intl.formatMessage(messages['library.header.settings.access'])}</Dropdown.Item>
```
**Deferrable:**

If I click on this menu item with View access, it loads the access page and then immediately kicks me back to the detail view.

I don't have official guidance from UX on this yet, but I imagine that for View-only users, we'll want to either (1) show the access page, even if it's read-only, (2) disable the User Access menu item, or (3) hide the User Access menu item.

Kelketek replied:
Added a comment to the README for now.

kdmccormick commented on src/library-authoring/library-access/LibraryAccessForm.jsx:
```diff
@@ -0,0 +1,194 @@
+                      <label className="h6 d-block" htmlFor="email">
+                        {intl.formatMessage(messages['library.access.form.email.label'])}
+                      </label>
+                      <Input
+                        name="email"
+                        id="email"
+                        type="text"
+                        placeholder={intl.formatMessage(messages['library.access.form.email.placeholder'])}
+                        value={data.email}
+                        onChange={onValueChange}
+                      />
```
**Deferrable:** The entered email disappears from the text input when there's a validation error. It strikes me as more user-friendly to leave it there for the user to edit.

Kelketek replied:
Fixed in latest push.

kdmccormick commented on src/library-authoring/library-access/UserAccessContainer.jsx:
```diff
@@ -0,0 +1,185 @@
+                    </Button>
+                  </Col>
+                </>
+              )}
```
Why is a Staff user able to elevate themselves to Admin level?

I was able to do confirm that it's possible locally, so apparently the API is fine with it as well. Is this existing behavior we brought over from V1 libraries :open_mouth: ?

bradenmacdonald replied:
*global* staff users can administer any course or library on the system; is that what you mean? https://github.com/edx/edx-platform/blob/8e1ea3e53637485cc016b1d9fd569263277ba362/openedx/core/djangoapps/content_libraries/permissions.py#L89-L91

People with "Author" or "Read-only" permission in this specific library definitely shouldn't be able to elevate their own permission.

Kelketek replied:
@bradenmacdonald is right here. I went back and tried granting staff access to honor@ and when logged in as them, I could not elevate my own privileges. I could only do so as a global staff user, which is how it should work.

kdmccormick replied:
Got it, thank you for clarifying that @Kelketek and @bradenmacdonald .

kdmccormick commented on src/library-authoring/common/data/constants.js:
```diff
@@ -16,6 +16,12 @@ export const SUBMISSION_STATUS = {
+  ADMIN: 'admin',
+  STAFF: 'author',
+  USER: 'read',
+};
```
**Note to self, no action required**:

It's weird that we make this mapping. To me, "Author" and "Read-Only" are much clearer than "Staff" and "User". I know that we're pulling this in from the V1 library editor; might be worth revisiting later.

bradenmacdonald replied:
If we can just change the UI to say Admin/Author/Read-Only now, let's do it. I think it's a source of confusion throughout the codebase when frontend terms don't match backend terms (e.g. `chapter` vs `section`), and it's hard to change once they're established. Here we have a bit of a greenfield, so let's use that to clean up and champion consistency.

Kelketek replied:
Done.

kdmccormick reviewed (changes requested):
Great work @Kelketek . I only have one significant question.

As for the comments of mine marked **Deferrable**, please either address them in this PR or simply add a note to the "Known Issues" section of the README. I would like to keep track of these type of issues as we go, but I don't want them to hold up merging, especially given that our UX team is likely to request changes to the interface as some point anyway.

Kelketek committed:
Fix some rougher edges.

Kelketek commented:
Closing this now that the edx PR is up to date with master.
